tests/spec_test_runner.go

```diff
@@ -0,0 +1,197 @@
+			}
+		case "assert_return", "action":
+			localVM := vm
+			if cmd.Action.Module != "" {
+				if target, ok := namedVMs[cmd.Action.Module]; ok {
+					localVM = target
+				} else {
+					panic("named module not found")
+				}
+			}
+			if localVM == nil {
+				panic("module not found")
+			}
+
+			switch cmd.Action.Type {
+			case "invoke":
+				entryID, ok := localVM.GetExportIndex(cmd.Action.Field)
+				if !ok {
+					panic("export not found (func)")
+				}
+				args := make([]uint64, 0)
+				for _, arg := range cmd.Action.Args {
+					var val uint64
+					fmt.Sscanf(arg.Value, "%d", &val)
+					args = append(args, val)
+				}
+				ret, err := localVM.ExecCode(int64(entryID), args...)
+				if err != nil {
+					panic(err)
+				}
+				if len(cmd.Expected) != 0 {
+					var _exp uint64
+					fmt.Sscanf(cmd.Expected[0].Value, "%d", &_exp)
+					exp := int64(_exp)
+					var result int64
+					if cmd.Expected[0].Type == "i32" || cmd.Expected[0].Type == "f32" {
+						result = int64(ret.(uint32))
+						exp = int64(uint32(exp))
+					} else {
+						result = int64(ret.(uint64))
+					}
+					if result != exp {
+						panic(fmt.Errorf("l%d: %s, ret mismatch: got %d, expected %d", cmd.Line, cfgPath, result, exp))
+					}
+				}
+			case "get":
+				globalID, ok := localVM.GetExportIndex(cmd.Action.Field)
+				if !ok {
+					panic("export not found (global)")
+				}
+				val, ok := localVM.GetGlobal(globalID)
+				if !ok {
+					panic("export not found (global)")
+				}
+				var exp uint64
+				fmt.Sscanf(cmd.Expected[0].Value, "%d", &exp)
+				if cmd.Expected[0].Type == "i32" || cmd.Expected[0].Type == "f32" {
+					val = uint64(uint32(val))
+					exp = uint64(uint32(exp))
+				}
+				if val != exp {
+					panic(fmt.Errorf("val mismatch: got %d, expected %d\n", val, exp))
+				}
+			default:
+				panic(cmd.Action.Type)
+			}
+		case "assert_trap":
+			localVM := vm
+			if cmd.Action.Module != "" {
+				if target, ok := namedVMs[cmd.Action.Module]; ok {
+					localVM = target
+				} else {
+					panic("named module not found")
+				}
+			}
+			if localVM == nil {
+				panic("module not found")
+			}
+			switch cmd.Action.Type {
+			case "invoke":
+				entryID, ok := localVM.GetExportIndex(cmd.Action.Field)
+				if !ok {
+					panic("export not found (func)")
+				}
+				args := make([]uint64, 0)
+				for _, arg := range cmd.Action.Args {
+					var val uint64
+					fmt.Sscanf(arg.Value, "%d", &val)
+					args = append(args, val)
+				}
+				_, err := localVM.ExecCode(int64(entryID), args...)
+				if err == nil {
+					panic(fmt.Errorf("L%d: %s, expect a trap\n", cmd.Line, cfgPath))
+				}
+			default:
+				panic(cmd.Action.Type)
+			}
+
+		case "assert_malformed", "assert_invalid", "assert_exhaustion", "assert_unlinkable",
+			"assert_return_canonical_nan", "assert_return_arithmetic_nan":
+			fmt.Printf("skipping %s\n", cmd.Type)
+		default:
+			panic(cmd.Type)
+		}
+		fmt.Printf("PASS L%d: %s\n", cmd.Line, cfgPath)
+	}
+}
+
+func main() {
```

Comment:
couldn't we pre-generate the `.json` files (via a `go generate` command) and make this run as part of the usual `testing.T` infrastructure?